server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java

```diff
@@ -316,78 +430,73 @@ protected LeafBucketCollector getLeafCollector(AggregationExecutionContext aggCt
+                            collectBucket(sub, doc, bucketOrd(bucket, topMatch.filterOrd));
+                            matched = true;
                         }
                     }
-
-                    if (otherBucketKey != null && false == matched) {
-                        collectBucket(sub, doc, bucketOrd(bucket, numFilters));
-                    }
                 }
+            }
 
-                @Override
-                public DocIdSetIterator competitiveIterator() throws IOException {
-                    if (useCompetitiveIterator) {
-                        // A DocIdSetIterator view of the filterIterators heap
-                        return new DisjunctionDISIApproximation(filterIterators);
-                    } else {
-                        return null;
-                    }
-                }
-            };
+            if (hasOtherBucket && false == matched) {
+                collectBucket(sub, doc, bucketOrd(bucket, numFilters));
+            }
         }
 
-        final long bucketOrd(long owningBucketOrdinal, int filterOrd) {
-            return owningBucketOrdinal * totalNumKeys + filterOrd;
+        @Override
+        public DocIdSetIterator competitiveIterator() throws IOException {
+            if (usesCompetitiveIterator) {
+                // A DocIdSetIterator view of the filterIterators heap
+                return new DisjunctionDISIApproximation(filterIterators);
+            }
+            return null;
         }
+    }
 
-        private static class FilterMatchingDisiWrapper extends DisiWrapper {
-            final int filterOrd;
+    private static class FilterMatchingDisiWrapper extends DisiWrapper {
+        final int filterOrd;
 
-            // Tracks the last doc that matches the filter.
-            int lastMatchingDoc = -1;
-            // Tracks the last doc that was checked for filter matching.
-            int lastCheckedDoc = -1;
+        // Tracks the last doc that matches the filter.
+        int lastMatchingDoc = -1;
+        // Tracks the last doc that was checked for filter matching.
+        int lastCheckedDoc = -1;
 
-            FilterMatchingDisiWrapper(Scorer scorer, int ord) {
-                super(scorer);
-                this.filterOrd = ord;
+        FilterMatchingDisiWrapper(Scorer scorer, int ord) {
+            super(scorer);
+            this.filterOrd = ord;
+        }
+
+        boolean checkDocForMatch(int doc) throws IOException {
+            // We need to cache the result of twoPhaseView.matches() since it's illegal to call it multiple times on the
+            // same doc, yet LeafBucketCollector#collect may be called multiple times with the same doc and multiple
+            // buckets.
+            if (lastCheckedDoc < doc) {
+                lastCheckedDoc = doc;
+                if (twoPhaseView == null || twoPhaseView.matches()) {
+                    lastMatchingDoc = doc;
+                }
             }
+            return (lastMatchingDoc == doc);
```

Comment:
Thinking out loud, we could change this logic to have a single comparison in the (not uncommon) case when `twoPhaseView` is `null`:

```java
boolean checkDocForMatch(int doc) throws IOException {
  if (twoPhaseView == null) {
    return true;
  }
  if (lastCheckedTwoPhaseDoc < doc) {
    lastCheckedTwoPhaseDoc = doc;
    if (twoPhaseView.matches()) {
      lastMatchingTwoPhaseDoc = doc;
    }
  }
  return lastMatchingTwoPhaseDoc == doc;
}
```

Reply:
Nice, I missed this. I created separate subclasses to avoid the branch - Martijn suggested this is hot code.